proposals/jvm-expose-boxed.md

```diff
@@ -0,0 +1,268 @@
+  * [Other design choices](#other-design-choices-1)
+  * [Further problems with reflection](#further-problems-with-reflection)
+* [Discarded potential features](#discarded-potential-features)
+* [Comparison with Scala](#comparison-with-scala)
+
+## Motivation
+
+[Value (or inline) classes](https://github.com/Kotlin/KEEP/blob/master/proposals/inline-classes.md) are an important tool to create a _lightweight wrapper_ over another class. One important example is given by classes that introduce additional invariants over another one, like an integer always greater than zero.
+
+```kotlin
+@JvmInline value class PositiveInt(val number: Int) {
+  init { require(number >= 0) }
+}
+```
+
+In order to keep its lightweight nature, the Kotlin compiler tries to use the _unboxed_ variant — in the case above, an integer itself — instead of its _boxed_ variant — the `PositiveInt` class — whenever possible. To prevent clashes between different overloads, the names of some of those functions must be _mangled_, that is, transformed in a specific way. The mangled functions operate directly on the underlying unboxed representation, with additional checks coming from the `init` block. The [inline classes](https://github.com/Kotlin/KEEP/blob/master/proposals/inline-classes.md) KEEP defines the whole compilation scheme, for this example is enough to know that `constructor-impl` is where the `init` block ends up living.
+
+```kotlin
+fun PositiveInt.add(other: PositiveInt): PositiveInt =
+  PositiveInt(this.number + other.number)
+
+// is compiled down to code similar to
+class PositiveInt {
+  // static member from the JVM perspective
+  public static constructor-impl(number: Int): Int {
+    require(number >= 0)
+    return number
+  }
+}
+// 1bc5 is completely made up
+fun Int.add-1bc5(other: Int): Int =
+  PositiveInt.constructor-impl(this.number + other.number)
+```
+
+The main drawback of this compilation scheme is that a type like `PositiveInt` cannot be (easily) consumed from Java (or other JVM languages other than Kotlin). This KEEP tries to rectify this problem, providing a way to consume value classes easily from Java. As a side benefit, library authors no longer need to choose between using a value class (which is more performant, but only available for Kotlin consumers) or a (data) class (which always requires boxing but can be used anywhere).
+
+> [!NOTE]
+> In the following code examples we use `static` to highlight those places where the compilation scheme produced a static method in JVM bytecode.
+> This does not imply any particular stance on future inclusion of `static` members in the Kotlin language.
+
+### Design goals
+
+"Exposing" a `value` class is not a single concept. The following are some scenarios that we would like to support:
+
+1. Creating boxed values: for example, creating an actual instance of `PositiveInt`.
+2. Calling operations that take or return value classes using their boxed representations.
```

Comment:
Besides calling, there's also overriding, which the current proposal doesn't handle, but perhaps it should be mentioned explicitly.